### PR TITLE
[Behat] Notification checker refactoring

### PR DESCRIPTION
**Pull request by Arminek**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | #4519 
| License         | MIT

### CODE REVIEW

File: src/Sylius/Behat/Service/Accessor/NotificationAccessor.php [hunk 1]
```diff
@@ -39,12 +39,25 @@ public function __construct(Session $session)
...
     public function hasSuccessMessage()
     {
-        $messageElement = $this->getMessageElement();
-        if (null === $messageElement) {
+        $messageType = $this->getMessageType();
+        if ('' === $messageType) {
+            return false;
+        }
+
+        return 'success' === $messageType;
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Service/Accessor/NotificationAccessor.php]: Isn't it enough?

Comment (michalmarcinkowski) [inline src/Sylius/Behat/Service/Accessor/NotificationAccessor.php]: It should be 😄

File: src/Sylius/Behat/Service/Accessor/NotificationAccessor.php [hunk 2]
```diff
@@ -91,14 +79,20 @@ public function getMessageType()
...
     /**
      * @return NodeElement
      */
     private function getMessageElement()
     {
-        return $this->session->getPage()->find('css', self::NOTIFICATION_ELEMENT_CSS);
+        $messageElement = $this->session->getPage()->find('css', self::NOTIFICATION_ELEMENT_CSS);
+
+        if (null === $messageElement) {
+            throw new \RuntimeException('There is no message element');
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Service/Accessor/NotificationAccessor.php]: Why not `NotFoundException`?

File: src/Sylius/Behat/Service/Accessor/NotificationAccessor.php [hunk 3]
```diff
@@ -91,14 +79,20 @@ public function getMessageType()
             return 'failure';
         }
 
-        return '';
+        throw new \RuntimeException('Cannot resolve message type');
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Service/Accessor/NotificationAccessor.php]: Why `RuntimeException`? These exceptions should be thrown only when something really bad happens, I would go with some standard exception.

File: src/Sylius/Behat/Exception/NotificationExpectationMismatchException.php [hunk 4]
```diff
@@ -11,33 +11,35 @@
...
      */
     public function __construct(
-        $missingNotificationType,
-        $missingNotification,
-        $foundedNotificationType,
-        $foundedNotification,
-        $code = 0, 
+        NotificationType $missingType,
+        $missingMessage,
+        NotificationType $foundedType,
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Exception/NotificationExpectationMismatchException.php]: foundMessage, here and everywhere else.

File: src/Sylius/Behat/spec/Service/NotificationTypeSpec.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+namespace spec\Sylius\Behat\Service;
+
+use PhpSpec\ObjectBehavior;
+
+/**
+ * @author Arkadiusz Krakowiak <arkadiusz.krakowiak@lakion.com>
+ */
+class NotificationTypeSpec extends ObjectBehavior
+{
+    function it_initialize_with_success_value()
```
Comment (pamil) [inline src/Sylius/Behat/spec/Service/NotificationTypeSpec.php]: `it_is_initialized_...`

File: src/Sylius/Behat/Service/NotificationChecker.php [hunk 6]
```diff
@@ -66,29 +67,47 @@ public function checkEditionNotification($resource)
...
     }
 
     /**
+     * @param NotificationType $type
+     *
+     * @return bool
+     */
+    private function hasType(NotificationType $type)
+    {
+        return (string) $type === (string) $this->notificationAccessor->getType();
```
Comment (pamil) [inline src/Sylius/Behat/Service/NotificationChecker.php]: That's quite a smell. Can `NotificationType` implement the flyweight pattern along with the existing static factory method, so every success notification type will share the same instance? Then we will be able to compare two types by `$type === $this->notificationAccessor->getType()`.

File: src/Sylius/Behat/Service/NotificationType.php [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+     * @var string
+     */
+    private $value;
+
+    /**
+     * @param string $value
+     */
+    private function __construct($value)
+    {
+        $this->value = $value;
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Service/NotificationType.php]: This class is a simple string container it doesn't make any sense unless we at least validate the possible values.

Comment (pamil) [inline src/Sylius/Behat/Service/NotificationType.php]: We do, it's private :)

File: src/Sylius/Behat/Service/NotificationType.php [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Behat\Service;
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Service/NotificationType.php]: This class does not fit `Service`...

File: src/Sylius/Behat/Exception/NotificationExpectationMismatchException.php (hunk 4 above)
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Exception/NotificationExpectationMismatchException.php]: I would rename this parameters to:
* `$expectedType`, `$expectedMessage`
* `$actualType`, `$actualMessage`

File: src/Sylius/Behat/Service/Accessor/NotificationAccessorInterface.php [hunk 9]
```diff
@@ -11,30 +11,20 @@
...
      * @return string
      */
     public function getMessage();
 
     /**
-     * @return string
+     * @return NotificationType
      */
-    public function getMessageType();
+    public function getType();
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Service/Accessor/NotificationAccessorInterface.php]: Why renaming this method?

File: src/Sylius/Behat/Service/NotificationChecker.php (hunk 6 above)
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Service/NotificationChecker.php]: Please don't do over engineering on such a simple thing, string comparison is fine for now :)

File: src/Sylius/Behat/Service/Accessor/NotificationAccessorInterface.php (hunk 9 above)
Comment (pamil) [inline src/Sylius/Behat/Service/Accessor/NotificationAccessorInterface.php]: IMO to be more consistent. `NotificationAccessorInterface::getMessage()` returns the notification message and `NotificationAccessorInterface::getType()` returns the notification type.

File: src/Sylius/Behat/Service/NotificationChecker.php (hunk 6 above)
Comment (pamil) [inline src/Sylius/Behat/Service/NotificationChecker.php]: I'm just predicting the WTF moments happening if someone tries to compare types without casting to a string :water_buffalo:

Comment (michalmarcinkowski) [inline src/Sylius/Behat/Service/NotificationChecker.php]: I believe `$type == $this->notificationAccessor->getType();` will work 😄

Comment (pamil) [inline src/Sylius/Behat/Service/NotificationChecker.php]: Please, do not use loose comparisons in PHP :crying_cat_face:

**Comment by pjedrzejewski** [comment]
Thanks Arek, we will iterate on that, happy rebasing folks!

**Pull request merged**